designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala

```diff
@@ -1,71 +1,61 @@
 import pl.touk.nussknacker.engine.canonicalgraph.CanonicalProcess
-import pl.touk.nussknacker.engine.deployment.{DeploymentData, DeploymentId, User}
+import pl.touk.nussknacker.engine.deployment.{DeploymentData, DeploymentId, ExternalDeploymentId, User}
 import pl.touk.nussknacker.restmodel.process.{ProcessIdWithName, ProcessingType}
-import pl.touk.nussknacker.restmodel.processdetails.BaseProcessDetails
+import pl.touk.nussknacker.restmodel.processdetails.{BaseProcessDetails, ProcessAction, ProcessShapeFetchStrategy}
 import pl.touk.nussknacker.ui.api.ListenerApiUser
-import pl.touk.nussknacker.ui.db.entity.ProcessActionEntityData
+import pl.touk.nussknacker.ui.db.entity.ProcessActionId
 import pl.touk.nussknacker.ui.listener.ProcessChangeEvent.{OnDeployActionFailed, OnDeployActionSuccess, OnFinished}
 import pl.touk.nussknacker.ui.listener.{ProcessChangeListener, User => ListenerUser}
-import pl.touk.nussknacker.ui.process.deployment.DeploymentActionType.{Cancel, Deployment}
 import pl.touk.nussknacker.ui.process.deployment.LoggedUserConversions.LoggedUserOps
+import pl.touk.nussknacker.ui.process.exception.{DeployingInvalidScenarioError, ProcessIllegalAction}
 import pl.touk.nussknacker.ui.process.repository.FetchingProcessRepository.FetchProcessesDetailsQuery
 import pl.touk.nussknacker.ui.process.repository.ProcessDBQueryRepository.ProcessNotFoundError
-import pl.touk.nussknacker.ui.process.repository.{DbProcessActionRepository, DeploymentComment, FetchingProcessRepository}
+import pl.touk.nussknacker.ui.process.repository.{DbProcessActionRepository, DeploymentComment, FetchingProcessRepository, ProcessDBQueryRepository, RepositoryManager}
 import pl.touk.nussknacker.ui.security.api.{LoggedUser, NussknackerInternalUser}
+import pl.touk.nussknacker.ui.validation.ProcessValidation
+import slick.dbio.DBIOAction
 
-import scala.concurrent.{ExecutionContext, Future}
+import java.time.Instant
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.concurrent.duration._
 import scala.util.control.NonFatal
 import scala.util.{Failure, Success}
 
+// Responsibility of this class is to wrap communication with DeploymentManager with persistent, transactional context.
+// It ensures that all actions are done consistently: do validations and ensures that only allowed actions
+// will be executed in given state. It sends notifications about finished actions.
+// Also thanks to it we are able to check if state on remote engine is the same as persisted state.
 class DeploymentServiceImpl(dispatcher: DeploymentManagerDispatcher,
-                            processRepository: FetchingProcessRepository[Future],
-                            actionRepository: DbProcessActionRepository,
+                            processRepository: FetchingProcessRepository[DB],
+                            actionRepository: DbProcessActionRepository[DB],
+                            dbioRunner: RepositoryManager,
```

Comment:
WDYT about changing the name of RepositoryManager? I left a TODO there. IMO we should avoid names like *Manager. Otherwise this RepositoryManager will be responsible for everything. If it manage repositories, why it doesn't return them, doesn't allow to add/remove new?

Reply:
Fine by me :)

Reply:
changed